File: web/src/api/watchlist.ts
import { apiFetch } from "./client";
import type { EvidenceSeverity } from "./filingMonitor";

/**
 * One Watchlist — a `collection` row with `collection_type=watchlist`
 * (ADR-016; backend/app/schemas/watchlist.py). Shares its underlying table
 * with Research Universes but is a distinct product concept: an analyst's
 * personal "which issuers do I care about" list.
 */
export interface WatchlistSummary {
  id: string;
  slug: string;
  name: string;
  description: string;
  issuer_count: number;
  issuers_with_new_developments: number;
  high_severity_count: number;
  last_activity_at: string | null;
  created_at: string;
  updated_at: string;
  /** Only populated when the request supplied `issuerId` (see `fetchWatchlists`). */
  contains_issuer: boolean | null;
}

export interface WatchlistListResponse {
  watchlists: WatchlistSummary[];
}

/** One watched issuer, with latest-development context — entirely assembled
 * from already-persisted alerts/memberships/securities, never new
 * intelligence the Watchlist itself creates. */
export interface WatchlistIssuerRow {
  issuer_id: string;
  issuer_legal_name: string;
  issuer_ticker: string | null;
  current_status: string[];
  latest_development_headline: string | null;
  latest_development_date: string | null;
  severity: EvidenceSeverity | null;
  new_developments_count: number;
  securities_count: number;
  rationale: string;
  added_at: string;
}

export interface WatchlistDetailResponse {
  watchlist: WatchlistSummary;
  issuers: WatchlistIssuerRow[];
}

export async function fetchWatchlists(issuerId?: string): Promise<WatchlistListResponse> {
  const params = new URLSearchParams();
  if (issuerId) params.set("issuer_id", issuerId);
  const query = params.toString();
  return apiFetch<WatchlistListResponse>(`/api/watchlists${query ? `?${query}` : ""}`);
}

export async function fetchWatchlist(watchlistId: string): Promise<WatchlistDetailResponse> {
  return apiFetch<WatchlistDetailResponse>(`/api/watchlists/${watchlistId}`);
}

export async function createWatchlist(input: {
  name: string;
  description?: string;
}): Promise<WatchlistSummary> {
  return apiFetch<WatchlistSummary>("/api/watchlists", {
    method: "POST",
    body: JSON.stringify({ name: input.name, description: input.description ?? "" }),
  });
}

export async function updateWatchlist(
  watchlistId: string,
  input: { name?: string; description?: string },
): Promise<WatchlistSummary> {
  return apiFetch<WatchlistSummary>(`/api/watchlists/${watchlistId}`, {
    method: "PATCH",
    body: JSON.stringify(input),
  });
}

export async function deleteWatchlist(watchlistId: string): Promise<void> {
  await apiFetch<void>(`/api/watchlists/${watchlistId}`, { method: "DELETE" });
}

export async function addIssuerToWatchlist(
  watchlistId: string,
  input: { issuer_id: string; rationale?: string },
): Promise<WatchlistSummary> {
  return apiFetch<WatchlistSummary>(`/api/watchlists/${watchlistId}/issuers`, {
    method: "POST",
    body: JSON.stringify({
      issuer_id: input.issuer_id,
      rationale: input.rationale ?? "Added to watchlist",
    }),
  });
}

export async function removeIssuerFromWatchlist(
  watchlistId: string,
  issuerId: string,
): Promise<void> {
  await apiFetch<void>(`/api/watchlists/${watchlistId}/issuers/${issuerId}`, {
    method: "DELETE",
  });
}

File: backend/app/schemas/watchlist.py
"""Response/request schemas for the Watchlists API (PLAN.md section 24.1;
Milestone 8).

Watchlists are `collection` rows with `collection_type=watchlist` (ADR-016)
— an analyst's personal "which issuers do I care about" list, distinct from
the organization-curated Research Universes that share the same table. This
module's job is purely presentation/aggregation of already-persisted data
(alerts, memberships, securities): no new intelligence is created here, and
building these responses makes zero Anthropic calls.
"""

from __future__ import annotations

from datetime import date, datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.core.types import EvidenceSeverity


class WatchlistSummary(BaseModel):
    """One Watchlist, with the roll-up counts an analyst dashboard needs —
    the row shape for the Watchlists landing page."""

    model_config = ConfigDict(frozen=True)

    id: UUID
    slug: str
    name: str
    description: str
    issuer_count: int
    issuers_with_new_developments: int
    high_severity_count: int
    last_activity_at: datetime | None
    created_at: datetime
    updated_at: datetime
    # Only populated when the request supplied `issuer_id` — lets the
    # "Add to Watchlist" component show already-added state without a
    # second round trip per watchlist.
    contains_issuer: bool | None = None


class WatchlistListResponse(BaseModel):
    model_config = ConfigDict(frozen=True)

    watchlists: list[WatchlistSummary]


class WatchlistCreateRequest(BaseModel):
    model_config = ConfigDict(frozen=True)

    name: str = Field(min_length=1, max_length=200)
    description: str = ""


class WatchlistUpdateRequest(BaseModel):
    model_config = ConfigDict(frozen=True)

    name: str | None = Field(default=None, min_length=1, max_length=200)
    description: str | None = None


class WatchlistIssuerRow(BaseModel):
    """One watched issuer, with the latest-development context Phase 4
    requires — assembled entirely from existing alert/membership/security
    data, never a new fact the Watchlist itself asserts."""

    model_config = ConfigDict(frozen=True)

    issuer_id: UUID
    issuer_legal_name: str
    issuer_ticker: str | None
    current_status: list[str]
    latest_development_headline: str | None
    latest_development_date: date | None
    severity: EvidenceSeverity | None
    new_developments_count: int
    securities_count: int
    rationale: str
    added_at: datetime


class WatchlistDetailResponse(BaseModel):
    model_config = ConfigDict(frozen=True)

    watchlist: WatchlistSummary
    issuers: list[WatchlistIssuerRow]


class AddIssuerToWatchlistRequest(BaseModel):
    model_config = ConfigDict(frozen=True)

    issuer_id: UUID
    rationale: str = "Added to watchlist"
